server/src/main/java/org/elasticsearch/cluster/metadata/MetadataAutoshardIndexService.java

```diff
@@ -0,0 +1,171 @@
+import org.elasticsearch.common.util.Maps;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.core.SuppressForbidden;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.index.Index;
+import org.elasticsearch.indices.IndicesService;
+import org.elasticsearch.indices.InvalidIndexNameException;
+import org.elasticsearch.logging.LogManager;
+import org.elasticsearch.logging.Logger;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.util.Map;
+
+public class MetadataAutoshardIndexService {
+    private static final Logger logger = LogManager.getLogger(MetadataAutoshardIndexService.class);
+
+    private final Settings settings;
+    private final ClusterService clusterService;
+    private final IndicesService indicesService;
+    private final AllocationService allocationService;
+    private final ThreadPool threadPool;
+
+    public MetadataAutoshardIndexService(
+        final Settings settings,
+        final ClusterService clusterService,
+        final IndicesService indicesService,
+        final AllocationService allocationService,
+        final ThreadPool threadPool
+    ) {
+        this.settings = settings;
+        this.clusterService = clusterService;
+        this.indicesService = indicesService;
+        this.allocationService = allocationService;
+        this.threadPool = threadPool;
+    }
+
+    public static void validateIndexName(String index, Metadata metadata, RoutingTable routingTable) {
+        if (routingTable.hasIndex(index) == false) {
+            throw new InvalidIndexNameException(index, "index does not exist");
+        }
+        /* Throw and error for datastream and system indexes.
+         * Datastream indices are autosharded using a different code path.
+         */
+    }
+
+    public void autoshardIndex(
+        final TimeValue masterNodeTimeout,
+        final TimeValue ackTimeout,
+        @Nullable final TimeValue waitForActiveShardsTimeout,
+        final AutoshardIndexClusterStateUpdateRequest request,
+        final ActionListener<ShardsAcknowledgedResponse> listener
+    ) {
+        logger.trace("autoshardIndex[{}]", request);
+        onlyAutoshardIndex(masterNodeTimeout, ackTimeout, request, listener.delegateFailureAndWrap((delegate, response) -> {
+            if (response.isAcknowledged()) {
+                logger.trace(
+                    "[{}] index autoshard acknowledged, waiting for active shards [{}]",
+                    request.index(),
+                    request.waitForActiveShards()
+                );
+                ActiveShardsObserver.waitForActiveShards(
+                    clusterService,
+                    new String[] { request.index() },
+                    request.waitForActiveShards(),
+                    waitForActiveShardsTimeout,
+                    delegate.map(shardsAcknowledged -> {
+                        if (shardsAcknowledged == false) {
+                            logger.debug(
```

Comment:
Might be good for this to be higher than debug, it's I guess not going to be spammy.